### PR TITLE
[Feature] SDPO

### PR DESCRIPTION
**Pull request by vmoens**
## Summary

Implements Self-Distillation Policy Optimization (SDPO), a new on-policy RL algorithm for LLM post-training based on the paper "Reinforcement Learning via Self-Distillation" (Hübotter et al., 2025).

**Key insight**: Instead of relying on sparse scalar rewards like GRPO, SDPO uses the model itself as a self-teacher by conditioning it on rich feedback (runtime errors, successful solutions, etc.). This enables dense, per-token credit assignment.

### New components:

- **`SDPOLoss`** (`torchrl/objectives/llm/sdpo.py`): Main loss module with:
  - KL, reverse-KL, and Jensen-Shannon divergence options
  - Top-K logit distillation for memory efficiency (~90% reduction)
  - EMA teacher and trust-region regularization for training stability
  - Entropy bonus and KL penalty to reference policy

- **`AddFeedbackContext`** (`torchrl/envs/llm/transforms/feedback.py`): Transform for preparing self-teacher context with environment feedback

- **`BuildTeacherContext`**: Chat history-based context builder for multi-turn scenarios

### Core equation:

```
L_SDPO = sum_t KL(pi_theta(·|x, y_{<t}) || stopgrad(pi_theta(·|x, f, y_{<t})))
```

Where `f` is the rich feedback (environment output + optionally a successful rollout).

### Reference:
- Paper: https://arxiv.org/abs/2601.20802
- Code: https://github.com/lasgroup/SDPO

## Test plan

- [ ] Unit tests for SDPOLoss with different divergence types (KL, JS)
- [ ] Unit tests for top-K logit distillation
- [ ] Unit tests for EMA teacher regularization
- [ ] Unit tests for AddFeedbackContext transform
- [ ] Integration with existing LLM training infrastructure

### CODE REVIEW

**Comment by gtnv** [comment]
hey @vmoens, is this still in progress? if not, should we close it?
